File: Server~/src/tools/getInteractableAtPositionTool.ts
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'get_interactable_at_position';
const toolDescription =
  "Returns the GameObjects under a screen-space point in front-to-back order (up to 5), with " +
  "clickable / selectable / raycaster info for each. Coordinates default to top-left origin (matching " +
  "screenshot pixel coordinates); pass origin='bottom-left' for Unity's native convention. " +
  "Useful diagnostic before ui_click_gameobject — answers 'is this button reachable?' / 'what's covering it?'. " +
  "Requires Play Mode and an active EventSystem.";

const paramsSchema = z.object({
  x: z
    .number()
    .int()
    .min(0)
    .describe("Screen X coordinate, in pixels. Top-left origin by default."),
  y: z
    .number()
    .int()
    .min(0)
    .describe("Screen Y coordinate, in pixels. Top-left origin by default."),
  origin: z
    .enum(['top-left', 'bottom-left'])
    .optional()
    .default('top-left')
    .describe(
      "Coordinate origin convention. 'top-left' matches screenshots and most UI tooling (default). " +
      "'bottom-left' matches Unity's native Screen / Input space."
    ),
});

export function registerGetInteractableAtPositionTool(
  server: McpServer,
  mcpUnity: McpUnity,
  logger: Logger
) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(
  mcpUnity: McpUnity,
  params: z.infer<typeof paramsSchema>
): Promise<CallToolResult> {
  const response = await mcpUnity.sendRequest({
    method: toolName,
    params: {
      x: params.x,
      y: params.y,
      origin: params.origin ?? 'top-left',
    },
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || `Failed to query interactable at (${params.x}, ${params.y})`
    );
  }

  return {
    content: [
      {
        type: 'text',
        text: response.message,
      },
      {
        type: 'text',
        text: JSON.stringify(
          {
            screenSize: response.screenSize,
            queryPosition: response.queryPosition,
            hitCount: response.hitCount,
            hits: response.hits,
          },
          null,
          2
        ),
      },
    ],
  };
}

File: Editor/Tools/GetInteractableAtPositionTool.cs
using System.Collections.Generic;
using McpUnity.Unity;
using McpUnity.Utils;
using Newtonsoft.Json.Linq;
using UnityEngine;
using UnityEngine.EventSystems;
using UnityEngine.UI;

namespace McpUnity.Tools
{
    /// <summary>
    /// Diagnostic tool: given a screen-space point (top-left origin by default, matching screenshot
    /// coordinates), report what would be hit by an EventSystem raycast. Returns up to 5 hits in
    /// front-to-back order, each annotated with whether it's clickable and which raycaster found it.
    /// Useful for "is this button reachable?" / "what's covering this region?" diagnostics.
    /// </summary>
    public class GetInteractableAtPositionTool : McpToolBase
    {
        private const int MaxHitsReported = 5;

        public GetInteractableAtPositionTool()
        {
            Name = "get_interactable_at_position";
            Description =
                "Returns the GameObjects under a screen-space point in front-to-back order, with " +
                "clickable / selectable / raycaster info. Coordinates default to top-left origin (matching " +
                "screenshot pixel coords); pass origin='bottom-left' for Unity's native convention. " +
                "Requires Play Mode and an active EventSystem.";
        }

        public override JObject Execute(JObject parameters)
        {
            int? xRaw = parameters?["x"]?.ToObject<int?>();
            int? yRaw = parameters?["y"]?.ToObject<int?>();
            string origin = parameters?["origin"]?.ToObject<string>() ?? "top-left";

            if (!xRaw.HasValue || !yRaw.HasValue)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Missing required parameters: x and y must both be provided.",
                    "validation_error");
            }

            if (!Application.isPlaying)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "get_interactable_at_position requires Play Mode.",
                    "not_in_play_mode");
            }

            EventSystem eventSystem = EventSystem.current;
            if (eventSystem == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "No active EventSystem in the scene.",
                    "no_event_system");
            }

            int screenWidth = Screen.width;
            int screenHeight = Screen.height;

            // Convert input to Unity screen coordinates (bottom-left origin).
            float unityX = xRaw.Value;
            float unityY = origin.Equals("bottom-left", System.StringComparison.OrdinalIgnoreCase)
                ? yRaw.Value
                : screenHeight - yRaw.Value;

            var pointerData = new PointerEventData(eventSystem)
            {
                position = new Vector2(unityX, unityY),
                button = PointerEventData.InputButton.Left
            };

            var results = new List<RaycastResult>();
            eventSystem.RaycastAll(pointerData, results);

            var hits = new JArray();
            int reportCount = Mathf.Min(results.Count, MaxHitsReported);
            for (int i = 0; i < reportCount; i++)
            {
                RaycastResult r = results[i];
                if (r.gameObject == null) continue;

                hits.Add(BuildHitInfo(r, i));
            }

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = results.Count == 0
                    ? $"No raycaster hit at ({xRaw.Value}, {yRaw.Value}, origin={origin})."
                    : $"{results.Count} hit(s) at ({xRaw.Value}, {yRaw.Value}, origin={origin}); top: " +
                      GameObjectResolver.GetHierarchyPath(results[0].gameObject),
                ["screenSize"] = new JObject
                {
                    ["width"] = screenWidth,
                    ["height"] = screenHeight
                },
                ["queryPosition"] = new JObject
                {
                    ["x"] = xRaw.Value,
                    ["y"] = yRaw.Value,
                    ["origin"] = origin,
                    ["unityX"] = unityX,
                    ["unityY"] = unityY
                },
                ["hitCount"] = results.Count,
                ["hits"] = hits
            };
        }

        private static JObject BuildHitInfo(RaycastResult hit, int depth)
        {
            GameObject go = hit.gameObject;
            GameObject clickHandler = ExecuteEvents.GetEventHandler<IPointerClickHandler>(go);
            Selectable selectable = go.GetComponentInParent<Selectable>();

            var info = new JObject
            {
                ["depth"] = depth,
                ["path"] = GameObjectResolver.GetHierarchyPath(go),
                ["isClickable"] = clickHandler != null,
                ["clickHandlerPath"] = clickHandler != null
                    ? GameObjectResolver.GetHierarchyPath(clickHandler)
                    : null,
                ["selectablePath"] = selectable != null
                    ? GameObjectResolver.GetHierarchyPath(selectable.gameObject)
                    : null,
                ["selectableInteractable"] = selectable != null ? selectable.IsInteractable() : (bool?)null,
                ["raycaster"] = hit.module != null ? hit.module.GetType().Name : null,
                ["sortingOrder"] = hit.sortingOrder,
                ["distance"] = hit.distance
            };

            RectTransform rect = go.transform as RectTransform;
            if (rect != null)
            {
                Canvas canvas = rect.GetComponentInParent<Canvas>();
                Camera cam = canvas != null && canvas.renderMode != RenderMode.ScreenSpaceOverlay
                    ? canvas.worldCamera
                    : null;

                Vector3[] corners = new Vector3[4];
                rect.GetWorldCorners(corners);
                Vector2 bottomLeft = RectTransformUtility.WorldToScreenPoint(cam, corners[0]);
                Vector2 topRight = RectTransformUtility.WorldToScreenPoint(cam, corners[2]);

                info["screenRect"] = new JObject
                {
                    ["x"] = Mathf.Min(bottomLeft.x, topRight.x),
                    ["y"] = Mathf.Min(bottomLeft.y, topRight.y),
                    ["width"] = Mathf.Abs(topRight.x - bottomLeft.x),
                    ["height"] = Mathf.Abs(topRight.y - bottomLeft.y)
                };
            }

            return info;
        }
    }
}
